### PR TITLE
Feat/detect detection filter unique on ips

### PR DESCRIPTION
**Pull request by oferda4**
Add optional unique_on {src_ip|dst_ip} to detection_filter for distinct IP address counting within the seconds window. Features:

Runtime uses a hash table per threshold entry for tracking unique IP addresses (both IPv4 and IPv6).
Follows detection_filter semantics: alerting starts after the threshold (>count), not at it.
On window expiry, the window is reset and the current packet's IP is recorded as the first distinct of the new window. Validation:
unique_on src_ip/dst_ip works with any IP protocol (unlike port-based unique_on which requires tcp/udp/sctp). Memory management:
Hash table memory is bounded by detect.thresholds.memcap.
Reuses existing counters: bitmap_memuse and bitmap_alloc_fail. Refactoring:
Added ThresholdDistinctAdd helper to consolidate port/IP tracking logic and reduce code duplication. Tests:
C unit tests for parsing unique_on src_ip and dst_ip options.

Changes:
v2:
- Update documentation
- Increase unit tests coverage
- Document use of hash table

v3:
- Cleanup unit tests

v4:
- Added validation to reject duplicate unique_on options
- Fixed memory accounting in ThresholdDistinctReset: counters are now properly decremented if hash table re-init fails
- Hash table is now pre-sized to count buckets instead of a fixed 64

v5:
- Fixed use-after-free: HashTableAdd stores pointers without copying, so ThresholdDistinctAddIP now allocates owned copies of IP address bytes. A Free callback (ThresholdDistinctIPFree) is registered with the hash table to automatically free copies on reset/free.
- Handle HashTableAdd allocation failure by freeing the IP copy
- Simplified window expiry: te->current_count = 1 is now unconditional
- Added documentation on two-phase memory allocation model (init-time bucket array vs. runtime per-entry copies)

v6:
- Replace HashTable with pre-allocated flat buffer (Address *distinct_ip_buf) for IP deduplication — eliminates all per-packet SCMalloc calls
- Add 4 new unit tests: IPv4 basic + buffer-full boundary, duplicate suppression, window reset, dst_ip with track by_src
- Restrict port-protocol validation to unique_on src_port/dst_port only (not src_ip/dst_ip)

Previous PR: https://github.com/OISF/suricata/pull/15086

Ticket: https://redmine.openinfosecfoundation.org/issues/8250
SV_BRANCH=https://github.com/OISF/suricata-verify/pull/2924

### CODE REVIEW

File: src/detect-engine-threshold.c [hunk 1]
```diff
@@ -186,43 +190,75 @@ static void ThresholdDistinctInit(ThresholdEntry *te, const DetectThresholdData
...
+            if (ctx.thash != NULL) {
+                (void)SC_ATOMIC_ADD(ctx.thash->memuse, bitmap_size);
+            }
+            SC_ATOMIC_ADD(threshold_bitmap_memuse, bitmap_size);
+        }
     }
+    /* IP-based distinct counting uses a pre-allocated flat buffer of td->count Address slots.
+     * All memory is allocated here at init time; no per-packet allocations are needed.
+     * Lookup is a linear scan (O(count)); reset only zeroes current_count. */
+    else if (td->unique_on == DF_UNIQUE_SRC_IP || td->unique_on == DF_UNIQUE_DST_IP) {
```
Comment (catenacyber) [inline src/detect-engine-threshold.c:230]: Could this be just else without if ?

File: src/detect-engine-threshold.c [hunk 2]
```diff
@@ -240,6 +276,64 @@ static inline void ThresholdDistinctAddPort(ThresholdEntry *te, uint16_t port)
...
+        key_size = 4;
+    } else if (addr->family == AF_INET6) {
+        key_size = 16;
+    } else {
+        return;
+    }
+
+    /* Linear scan through valid entries. When the buffer is full, only
+     * max_count entries are stored; cap the scan range accordingly. */
+    const uint32_t scan_limit = (te->current_count < max_count) ? te->current_count : max_count;
```
Comment (catenacyber) [inline src/detect-engine-threshold.c:297]: Could always be `te->current_count` as `te->current_count > max_count` never happens, right ?

File: src/detect-detection-filter.c [hunk 3]
```diff
@@ -173,10 +173,21 @@ static DetectThresholdData *DetectDetectionFilterParse(const char *rawstr)
...
+                goto error;
+            }
+            if (strcasecmp(args[i], "src_port") == 0)
+                df->unique_on = DF_UNIQUE_SRC_PORT;
+            else if (strcasecmp(args[i], "dst_port") == 0)
+                df->unique_on = DF_UNIQUE_DST_PORT;
+            else if (strcasecmp(args[i], "src_ip") == 0)
+                df->unique_on = DF_UNIQUE_SRC_IP;
+            else
+                df->unique_on = DF_UNIQUE_DST_IP;
```
Comment (catenacyber) [inline src/detect-detection-filter.c:189]: Should we enforce a global limit that we do not want Suricata to count 1M distinct IPs ?

File: src/detect-engine-threshold.c [hunk 4]
```diff
@@ -240,6 +276,64 @@ static inline void ThresholdDistinctAddPort(ThresholdEntry *te, uint16_t port)
...
+        COPY_ADDRESS(addr, &te->distinct_ip_buf[te->current_count]);
+    }
+    te->current_count++;
+}
+
+/**
+ * \brief Try to add a distinct value (port or IP) based on unique_on type.
+ * \return true if distinct tracking is active (bitmap/hash allocated), false if fallback needed
+ */
+static inline bool ThresholdDistinctAdd(
```
Comment (catenacyber) [inline src/detect-engine-threshold.c:318]: This function now always returns true as we do not have anymore allocs problem

**Review by catenacyber — CHANGES_REQUESTED** [review]
Thanks for the work, I think one last round for nits should make it fine

CI : ✅ 
Code : some last details to polish
Commits segmentation : cool thanks
Commit messages : good thanks
Git ID set : ok
CLA : you already contributed
Doc update : cool
Redmine ticket : ok
Rustfmt : no rust
Tests : looks good
Dependencies added: none

File: src/detect-engine-threshold.c (hunk 2 above)
Comment (catenacyber) [reply src/detect-engine-threshold.c:297]: (Genuine question to understand the code, I am not sure)

File: src/detect-detection-filter.c (hunk 3 above)
Comment (catenacyber) [reply src/detect-detection-filter.c:189]: (Otherwise, we can go fancy, and have a hashtable whose elements point in the preallocated flat buffer)

File: src/detect-engine-threshold.c (hunk 2 above)
Comment (oferda4) [reply src/detect-engine-threshold.c:297]: `current_count` can exceed `max_count` - it's the true count of distinct IPs seen, while only max_count are stored. The scan_limit caps the scan to actually-stored entries. 
I can add a brief comment clarifying this.

File: src/detect-engine-threshold.c (hunk 4 above)
Comment (oferda4) [reply src/detect-engine-threshold.c:318]: The function can still return false when the bitmap/buffer wasn't allocated (memcap exceeded or alloc failure at init time). The callers use this to fall back to classic counting.

File: src/detect-detection-filter.c (hunk 3 above)
Comment (oferda4) [reply src/detect-detection-filter.c:189]: Good point - I'll add a cap in the parser that rejects rules with excessively large count when unique_on is src_ip/dst_ip. With the current linear scan, keeping it bounded to something reasonable (maybe 1000 or 10000?) avoids both the memory and the O(n) scan cost concern. If there's ever demand for higher counts, we can switch to a hash-backed buffer as a follow-up.
